File: packages/typia/src/factories/internal/metadata/iterate_metadata_array.ts
import ts from "typescript";

import { MetadataArray } from "../../../schemas/metadata/MetadataArray";
import { MetadataArrayType } from "../../../schemas/metadata/MetadataArrayType";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { emplace_metadata_array_type } from "./emplace_metadata_array_type";

export const iterate_metadata_array = (
  props: IMetadataIteratorProps,
): boolean => {
  const array: ts.Type | null =
    props.checker.isArrayType(props.type) === false
      ? find_array_extended({
          checker: props.checker,
          memory: new Map(),
          type: props.type,
        })
      : props.type;
  if (array === null) return false;

  const arrayType: MetadataArrayType = emplace_metadata_array_type({
    ...props,
    array,
  });
  ArrayUtil.add(
    props.metadata.arrays,
    MetadataArray.create({
      type: arrayType,
      tags: [],
    }),
    (elem) => elem.type.name === arrayType.name,
  );
  return true;
};

const find_array_extended = (props: {
  checker: ts.TypeChecker;
  memory: Map<ts.Type, ts.Type | null>;
  type: ts.Type;
}): ts.Type | null => {
  const cached = props.memory.get(props.type);
  if (cached !== undefined) return null;

  props.memory.set(props.type, null);
  const res: ts.Type | null = (() => {
    if (props.type.isClassOrInterface() === false) return null;
    for (const t of props.type.resolvedBaseTypes ?? [])
      if (props.checker.isArrayType(t)) return t;
      else {
        const res: ts.Type | null = find_array_extended({
          ...props,
          type: t,
        });
        if (res !== null) return res;
      }
    return null;
  })();
  props.memory.set(props.type, res);
  return res;
};

File: packages/core/src/factories/internal/metadata/iterate_metadata_array.go
package metadata

func IterateMetadataArray(props IMetadataIteratorProps) bool {
	array := FindArrayExtended(props.Checker, map[TypeLike]TypeLike{}, props.Type)
	if array == nil {
		return false
	}
	arrayType := EmplaceMetadataArrayType(props, array)
	addArray(&props.Metadata.Arrays, MetadataArray{Type: arrayType})
	return true
}

func FindArrayExtended(checker TypeChecker, memory map[TypeLike]TypeLike, typ TypeLike) TypeLike {
	if typ == nil {
		return nil
	}
	if cached, ok := memory[typ]; ok {
		return cached
	}
	memory[typ] = nil
	if typ.IsArray() {
		memory[typ] = typ
		return typ
	}
	for _, base := range typ.BaseTypes() {
		if found := FindArrayExtended(checker, memory, base); found != nil {
			memory[typ] = found
			return found
		}
	}
	return nil
}
